crates/nix-ninja-task/src/derived_file.rs

```diff
@@ -7,53 +7,85 @@ use std::path::PathBuf;
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct DerivedFile {
     pub path: SingleDerivedPath,
-    pub source: PathBuf,
+    pub build_path: PathBuf, // Where file appears in build dir (symlink destination)
+    pub rel_path: Option<PathBuf>, // Where file appears within derived path (None for opaque)
 }
 
 impl DerivedFile {
     pub fn to_encoded(&self) -> String {
+        let path_str = match &self.path {
+            SingleDerivedPath::Opaque(store_path) => {
+                store_path.path().to_string_lossy().to_string()
+            }
+            SingleDerivedPath::Built(built_path) => {
+                built_path.placeholder().to_string_lossy().to_string()
+            }
+        };
+        let rel_path_str = self
+            .rel_path
+            .as_ref()
+            .map(|p| p.to_string_lossy().to_string())
+            .unwrap_or_default();
         format!(
-            "{}:{}",
-            self.path.to_input().display(),
-            &self.source.to_string_lossy()
+            "{}:{}:{}",
+            path_str,
+            &self.build_path.to_string_lossy(),
+            rel_path_str
         )
     }
 
     pub fn from_encoded(encoded: &str) -> Result<Self> {
-        // Split by colon to separate path from source
         let parts: Vec<&str> = encoded.split(':').collect();
-        if parts.len() != 2 {
+        if parts.len() != 3 {
             return Err(anyhow!(
-                "Expected one ':' in encoded derived file but got {}",
+                "Expected two ':' in encoded derived file but got {}",
```

Comment:
You can use the in-place notation when you're interpolating a simple variable:
```suggestion
                "Expected two ':' in encoded derived file but got {encoded}",
```